### PR TITLE
feat(core): add snapshot for hippy

### PR DESCRIPTION
**Pull request by churchill-zhang**
# Pre-PR Checklist

- [x] I added/updated relevant documentation.
- [x] I followed the [Convention Commit](https://conventionalcommits.org/) guideline with maximum 72 characters to submit commit message.
- [x] I squashed the repeated code commits.
- [x] I signed the [CLA].
- [x] I added/updated test cases to check the change I am making.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by ilikethese — DISMISSED** [review]

**Review by ozonelmy — DISMISSED** [review]
LGTM

**Review by siguangli — APPROVED** [review]

**Review by medns — CHANGES_REQUESTED** [review]
changes needs to be made

**Review by zoomchan-cxj — CHANGES_REQUESTED** [review]
js 是否需要判断属性存在性

**Review by siguangli — APPROVED** [review]

**Review by zoomchan-cxj — APPROVED** [review]

**Review by ilikethese — APPROVED** [review]

**Review by siguangli — DISMISSED** [review]

**Review by siguangli — APPROVED** [review]

**Review by ilikethese — APPROVED** [review]

**Comment by churchill-zhang** [comment]
> changes needs to be made

已拆多个commit

**Comment by churchill-zhang** [comment]
> js 是否需要判断属性存在性

corejs相关改动已还原

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 1]
```diff
@@ -107,6 +111,8 @@ void HandleUncaughtJsError(v8::Local<v8::Message> message,
   TDF_BASE_DLOG(INFO) << "HandleUncaughtJsError end";
 }
 
+// REGISTER_EXTERNAL_REFERENCES(HandleUncaughtJsError)
```
Comment (medns) [inline android/sdk/src/main/jni/src/bridge/entry.cc:114]: unnecessary comment?

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 2]
```diff
@@ -169,6 +182,8 @@ void NativeCallback(const hippy::napi::CallbackInfo& info, void* data) {
   hippy::bridge::CallJava(info, runtime_id);
 }
 
+// REGISTER_EXTERNAL_REFERENCES(NativeCallback)
```
Comment (medns) [inline android/sdk/src/main/jni/src/bridge/entry.cc:185]: unnecessary comment?

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 3]
```diff
@@ -340,6 +354,82 @@ bool RunScriptInternal(const std::shared_ptr<Runtime>& runtime,
...
+      .count();
+  TDF_BASE_LOG(INFO) << "blob size = " << blob.raw_size << ", buffer size = " << snapshot_data.buffer_holder.size()
+    << ", cost = " << (time_end - time_begin);
+  auto snapshot_uri = JniUtils::ToStrView(j_env, j_snapshot_uri);
+  bool save_file_ret = HippyFile::SaveFile(snapshot_uri, snapshot_data.buffer_holder);
+  if (!save_file_ret) {
+    return static_cast<jint>(CreateSnapshotResult::kSaveSnapshotFailed);
+  }
+  vm = nullptr;
+  engine = nullptr;
```
Comment (medns) [inline android/sdk/src/main/jni/src/bridge/entry.cc:429]: unnecessary set to `nullptr`

File: core/include/core/vm/v8/snapshot_data.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#pragma clang diagnostic pop
+
+// Snapshot file layout:
+// magic number 0x66886688
+// sdk version string "2.15.7"
+// blob length uint32_t
+// blob raw data
+
+constexpr uint32_t kMagicNumber = 0x66886688;
+constexpr char kSdkVersion[] = "2.15.8";
```
Comment (medns) [inline core/include/core/vm/v8/snapshot_data.h:41]: SDK version may be imported from outside?

File: core/include/core/vm/v8/snapshot_data.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ public:
+  uint32_t magic_number;
+  std::string sdk_version;
+  v8::StartupData startup_data;
+
+  std::vector<uint8_t> buffer_holder; // hold v8::StartupData data
+  std::any external_buffer_holder;    // hold DirectBuffer to avoid copying
+
+  void WriteMetaData(v8::StartupData data);
+  bool ReadMetaData();                // use meta data in buffer_holder
```
Comment (medns) [inline core/include/core/vm/v8/snapshot_data.h:53]: `Metadata` is a proper noun, change `ReadMetaData` to `ReadMetadata`

File: core/include/core/vm/v8/snapshot_serializer.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+class SnapshotSerializer {
+ public:
+  SnapshotSerializer(std::vector<uint8_t>& buffer): buffer_(buffer) {}
+  void WriteUInt32(uint32_t value);
+  void WriteString(const std::string& value);
+  void WriteBuffer(const void* p, size_t length);
+
+ private:
+  std::vector<uint8_t>& buffer_;
+  uint32_t position_ = 0;
```
Comment (medns) [inline core/include/core/vm/v8/snapshot_serializer.h:38]: change `position_` to a  `size_t` type

File: core/src/base/file.cc [hunk 7]
```diff
@@ -30,24 +30,42 @@
...
-  if (file.is_open()) {
-    std::streamsize len;
-    bool is_success = numeric_cast<size_t, std::streamsize>(content.length(), len);
-    if (is_success) {
-      file.write(content.c_str(), len);
+using unicode_string_view = tdf::base::unicode_string_view;
+
+bool HippyFile::ReadFile(const unicode_string_view& file_path,
+                         const std::function<void*(size_t)>& realloc,
+                         bool is_auto_fill) {
```
Comment (medns) [inline core/src/base/file.cc:37]: why need auto fill?

File: core/src/vm/v8/snapshot_data.cc [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  startup_data = data;
+  SnapshotSerializer serializer(buffer_holder);
+  serializer.WriteUInt32(kMagicNumber);
+  serializer.WriteString(kSdkVersion);
+  auto size = hippy::base::checked_numeric_cast<int, uint32_t>(data.raw_size);
+  serializer.WriteUInt32(size);
+  serializer.WriteBuffer(data.data, size);
+}
+
+bool SnapshotData::ReadMetaData() {
```
Comment (medns) [inline core/src/vm/v8/snapshot_data.cc:39]: output detial messages(like a `read %s failed, excepted %s, received %s`) when an exception occurs

File: core/src/vm/v8/snapshot_serializer.cc [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  auto length = value.length();
+  WriteUInt32(hippy::base::checked_numeric_cast<size_t, uint32_t>(length));
+  WriteBuffer(value.c_str(), length);
+}
+
+void SnapshotSerializer::WriteBuffer(const void* p, size_t length) {
+  if (buffer_.capacity() <= buffer_.size() + length) {
+    buffer_.resize(buffer_.size() + length);
+  }
+  memcpy(&buffer_[0] + position_, p, length);
```
Comment (medns) [inline core/src/vm/v8/snapshot_serializer.cc:41]: use `std::copy_n` instead

**Review by medns — CHANGES_REQUESTED** [review]

File: core/src/base/file.cc (hunk 7 above)
Comment (churchill-zhang) [reply core/src/base/file.cc:37]: 读c风格字符串需要，避免外部拷贝

**Review by ilikethese — DISMISSED** [review]

**Review by siguangli — APPROVED** [review]

**Review by ilikethese — APPROVED** [review]

**Review by medns — APPROVED** [review]
LGTM

**Pull request merged**